File: examples/shell-integration.js
#!/usr/bin/env node

const { spawn, execSync } = require('child_process');
const fs = require('fs');

// Attempt 1: Basic spawn (aliases won't work)
function basicShellAttempt() {
    console.log("\n1. Basic spawn attempt (aliases won't work):");
    
    // Can't use -i (interactive) flag because:
    // - It would block waiting for input with no PTY
    // - Node's spawn doesn't allocate a pseudo-terminal
    // - Would block the event loop indefinitely
    // Must use -c to run command and exit immediately
    const shell = spawn('bash', ['-c', 'alias hello="echo Hello from alias"; hello']);
    
    shell.stdout.on('data', (data) => {
        console.log(`stdout: ${data}`);
    });
    
    shell.stderr.on('data', (data) => {
        console.error(`stderr: ${data}`);
    });
    
    shell.on('close', (code) => {
        console.log(`Shell exited with code ${code}`);
        console.log("Result: No output - aliases don't work in non-interactive mode!");
    });
}

// Attempt 2: Try to use FIFOs (problematic in Node)
async function fifoAttempt() {
    console.log("\n2. FIFO attempt:");
    
    const fifoPath = '/tmp/node_shell_fifo';
    
    // Clean up any existing FIFO first
    try {
        fs.unlinkSync(fifoPath);
    } catch (e) {
        // Ignore if doesn't exist
    }
    
    try {
        // Node can't create FIFOs directly - must shell out
        execSync(`mkfifo ${fifoPath}`);
        console.log("Created FIFO using execSync (not native!)");
        
        // This is where it gets problematic...
        // Reading from FIFO blocks the entire event loop!
        
        // Remember: In Python we used fork() to handle this cleanly:
        // - Parent process could block on FIFO read
        // - Child process wrote to FIFO
        // - No event loop to worry about!
        // In Node, we can't fork() and blocking kills everything
        
        // Spawn a child to write to FIFO
        const writer = spawn('bash', ['-c', `echo "Hello from FIFO" > ${fifoPath}`]);
        
        // Try to read - this will block everything!
        console.log("Warning: About to block event loop reading FIFO...");
        
        // We have to use readFileSync or it won't work at all
        setTimeout(() => {
            try {
                const data = fs.readFileSync(fifoPath, 'utf8');
                console.log(`Received: ${data}`);
            } catch (e) {
                console.error("Failed to read from FIFO:", e.message);
            }
        }, 100);
        
    } catch (e) {
        console.error("FIFO operations failed:", e.message);
    } finally {
        try {
            fs.unlinkSync(fifoPath);
        } catch (e) {}
    }
}

// Demonstrate signal handling limitations
function signalHandlingDemo() {
    console.log("\n3. Signal handling limitations:");
    
    // Spawn a shell that runs multiple processes
    const child = spawn('bash', ['-c', 'echo "Parent bash PID: $$"; sleep 5 & echo "Background sleep PID: $!"; wait']);
    
    child.stdout.on('data', (data) => {
        console.log(`Child output: ${data.toString().trim()}`);
    });
    
    // Try to kill after 1 second
    setTimeout(() => {
        console.log("Killing child process...");
        child.kill('SIGTERM');
        // Problem: This only kills bash, not the background sleep!
        console.log("Killed bash, but background processes may still be running!");
    }, 1000);
    
    console.log("Node can't:");
    console.log("- Create new sessions (no setsid)");
    console.log("- Kill entire process groups");
    console.log("- Properly manage child process trees");
}

// Main demo
async function main() {
    console.log("Node.js Shell Integration Demo");
    console.log("=".repeat(40));
    console.log("\nShowing why Node struggles with shell integration...");
    
    basicShellAttempt();
    
    setTimeout(async () => {
        await fifoAttempt();
        
        setTimeout(() => {
            signalHandlingDemo();
            
            console.log("\n" + "=".repeat(40));
            console.log("Summary of Node.js limitations:");
            console.log("- No native PTY support (need compiled addons)");
            console.log("- No native FIFO creation");
            console.log("- FIFOs block the event loop");
            console.log("- Limited signal handling");
            console.log("- Can't create proper interactive shells");
            console.log("\nResult: Aliases and proper shell integration are very difficult!");
        }, 2000);
    }, 1000);
}

main().catch(console.error);

File: examples/shell-integration.py
#!/usr/bin/env python3

import os
import pty
import select
import subprocess
import sys
from pathlib import Path

def demo_fifo_communication():
    """Demonstrate named pipe creation and usage"""
    fifo_path = Path('/tmp/shell_demo_fifo')
    
    # Clean up any existing FIFO
    if fifo_path.exists():
        fifo_path.unlink()
    
    # Create FIFO natively
    os.mkfifo(fifo_path)
    print(f"Created FIFO at {fifo_path}")
    
    # Fork to demonstrate IPC
    if os.fork() == 0:
        # Child: write to FIFO
        with open(fifo_path, 'w') as fifo:
            fifo.write("Hello from child process via FIFO!\n")
        sys.exit(0)
    else:
        # Parent: read from FIFO (this would block in Node!)
        with open(fifo_path, 'r') as fifo:
            message = fifo.read()
            print(f"Received via FIFO: {message.strip()}")
    
    # Cleanup
    fifo_path.unlink()

def create_interactive_shell():
    """Create a proper interactive shell with PTY that supports aliases"""
    
    # Create a pseudo-terminal
    # parent_fd: used by parent process to read/write to the terminal
    # child_fd: connected to stdin/stdout/stderr of the child process
    parent_fd, child_fd = pty.openpty()
    
    # Launch bash in interactive mode with proper environment
    proc = subprocess.Popen(
        ['bash', '-i'],  # -i for interactive (enables aliases!)
        stdin=child_fd,
        stdout=child_fd,
        stderr=child_fd,
        preexec_fn=os.setsid,  # setsid = "set session ID" - creates new session/process group
        env={**os.environ, 'PS1': '$ '}  # Simple prompt
    )
    
    # Close child FD in parent
    os.close(child_fd)
    
    return proc, parent_fd

def main():
    print("Python Shell Integration Demo")
    print("=" * 40)

    # 1. Demo FIFO communication
    print("\n1. Testing FIFO communication:")
    demo_fifo_communication()

    
    # 2. Create interactive shell with PTY
    print("\n2. Creating interactive shell with alias support:")
    proc, parent_fd = create_interactive_shell()
    
    try:
        # Send a command to create an alias
        os.write(parent_fd, b"alias hello='echo Hello from alias!'\n")
        os.write(parent_fd, b"hello\n")  # This will work!
        os.write(parent_fd, b"sleep 2\n")
        os.write(parent_fd, b"echo 'Direct command'\n")
        os.write(parent_fd, b"exit\n")
        
        # Read output with proper select() handling
        print("\nShell output:")
        print("-" * 40)
        
        while proc.poll() is None:
            # Use select for non-blocking I/O
            r, _, _ = select.select([parent_fd], [], [], 0.1)
            if r:
                output = os.read(parent_fd, 1024)
                print(output.decode('utf-8', errors='replace'), end='')
    
    finally:
        os.close(parent_fd)
        
        # Ensure process is terminated
        if proc.poll() is None:
            proc.terminate()
            proc.wait()
    
    print("\n" + "=" * 40)
    print("Demo complete! Aliases worked in the interactive shell.")

if __name__ == '__main__':
    main()
